api/src/test/java/org/openmrs/scheduler/tasks/AutoRetireUsersTaskTest.java

```diff
@@ -0,0 +1,146 @@
+		when(userService.getAllUsers()).thenReturn(Collections.singletonList(inactiveUser));
+
+		autoRetireUsersTask.execute();
+
+		verify(userService, atLeastOnce()).retireUser(inactiveUser, AutoRetireUsersTask.AUTO_RETIRE_REASON);
+	}
+
+	@Test
+	public void shouldNotRetireUsersWhoseInactivityDoNotExceedNumberOfDaysToRetire() {
+		administrationService.saveGlobalProperty(new GlobalProperty(OpenmrsConstants.GP_NUMBER_OF_DAYS_TO_AUTO_RETIRE_USERS, TWO_DAYS_PROPERTY_VALUE));
+
+		User activeUser = getDefaultUser();
+
+		when(userService.getLastLoginTime(any())).thenReturn(String.valueOf(System.currentTimeMillis() - ONE_DAY_IN_MILLISECONDS));
+		when(userService.getAllUsers()).thenReturn(Collections.singletonList(activeUser));
+
+		autoRetireUsersTask.execute();
+
+		verify(userService, never()).retireUser(activeUser, AutoRetireUsersTask.AUTO_RETIRE_REASON);
+	}
+
+	@Test
+	public void shouldNotRetireAlreadyRetiredUsers() {
+		administrationService.saveGlobalProperty(new GlobalProperty(OpenmrsConstants.GP_NUMBER_OF_DAYS_TO_AUTO_RETIRE_USERS, ONE_DAY_PROPERTY_VALUE));
+
+		User retiredUser = getDefaultUser();
+		retiredUser.setRetired(true);
+
+		when(userService.getAllUsers()).thenReturn(Collections.singletonList(retiredUser));
+
+		autoRetireUsersTask.execute();
+
+		verify(userService, never()).retireUser(retiredUser, AutoRetireUsersTask.AUTO_RETIRE_REASON);
+	}
+
+	@Test
+	public void shouldNotRetireUsersThatAreInactiveSinceCreationAndInactivityDoesNotExceedNumberOfDaysToRetire() {
+		administrationService.saveGlobalProperty(new GlobalProperty(OpenmrsConstants.GP_NUMBER_OF_DAYS_TO_AUTO_RETIRE_USERS, ONE_DAY_PROPERTY_VALUE));
+
+		User adminUser = getDefaultUser();
+		adminUser.setDateCreated(new Date(System.currentTimeMillis() - TWENTY_THREE_HOURS_IN_MILLISECONDS));
+
+		when(userService.getAllUsers()).thenReturn(Collections.singletonList(adminUser));
+
+		autoRetireUsersTask.execute();
+
+		verify(userService, never()).retireUser(adminUser, AutoRetireUsersTask.AUTO_RETIRE_REASON);
+	}
+
+	@Test
+	public void shouldRetireUsersThatAreInactiveSinceCreationAndInactivityExceedsNumberOfDaysToRetire() {
+		administrationService.saveGlobalProperty(new GlobalProperty(OpenmrsConstants.GP_NUMBER_OF_DAYS_TO_AUTO_RETIRE_USERS, ONE_DAY_PROPERTY_VALUE));
+
+		User adminUser = getDefaultUser();
+		adminUser.setDateCreated(new Date(System.currentTimeMillis() - TWO_DAYS_IN_MILLISECONDS));
+
+		when(userService.getAllUsers()).thenReturn(Collections.singletonList(adminUser));
+
+		autoRetireUsersTask.execute();
+
+		verify(userService, atLeastOnce()).retireUser(eq(adminUser), anyString());
+	}
+
+	@Test
+	public void shouldNotRetireSuperUsers() {
+		administrationService.saveGlobalProperty(new GlobalProperty(OpenmrsConstants.GP_NUMBER_OF_DAYS_TO_AUTO_RETIRE_USERS, ONE_DAY_PROPERTY_VALUE));
+
+		User adminUser = getDefaultUser();
+		adminUser.addRole(new Role(RoleConstants.SUPERUSER));
```

Comment:
Will this test fail if i remove the above line of adding the super user role?

Reply:
@dicksonmulli did you see the above comment?

Reply:
Commenting should be working now. 

Regarding the above concern, kindly refer to the other reply that I shared with you on slack.

Reply:
It is better to share the reply here. These are public comments that are usually read by other community members who may join the discussion.

Reply:
https://github.com/openmrs/openmrs-core/pull/4650#issuecomment-2168116350

Reply:
Ideally it should fail but the reason it won't fail when you remove this line is because the algorithm will check if the inactivity exceeds the number of days to retire, of which it doesn't, unless we add unnecessary stubbings of a last login time or created date - e.g. when(userService.getLastLoginTime(any())).thenReturn(String.valueOf(System.currentTimeMillis() - TWO_DAYS_IN_MILLISECONDS));
Note that these unnecessary stubbing would cause an error.

Reply:
I have not properly understood your response. Let me ask differently, if i removed the line which adds the super user role to the user, will this test fail?